go/libkb/per_user_key.go

```diff
@@ -0,0 +1,598 @@
+package libkb
+
+import (
+	"crypto/rand"
+	"crypto/subtle"
+	"encoding/base64"
+	"errors"
+	"fmt"
+	"sync"
+
+	keybase1 "github.com/keybase/client/go/protocol/keybase1"
+	"github.com/keybase/go-codec/codec"
+	"golang.org/x/crypto/nacl/secretbox"
+	context "golang.org/x/net/context"
+)
+
+const PerUserKeySeedSize = 32
+
+// A secretbox containg a seed encrypted for its successor generation
+type PerUserKeyPrev string
+
+type PerUserKeySeed [PerUserKeySeedSize]byte
+
+func (s *PerUserKeySeed) DeriveSigningKey() (*NaclSigningKeyPair, error) {
+	derived, err := DeriveNonReversed(*s, DeriveReasonPUKSigning)
+	if err != nil {
+		return nil, err
+	}
+	res, err := MakeNaclSigningKeyPairFromSecret(derived)
+	return &res, err
+}
+
+func (s *PerUserKeySeed) DeriveEncryptionKey() (*NaclDHKeyPair, error) {
+	derived, err := DeriveNonReversed(*s, DeriveReasonPUKEncryption)
+	if err != nil {
+		return nil, err
+	}
+	res, err := MakeNaclDHKeyPairFromSecret(derived)
+	return &res, err
+}
+
+// derivePrevKey derives the symmetric key used to secretbox the previous generation seed.
+func (s *PerUserKeySeed) derivePrevKey() (res NaclSecretBoxKey, err error) {
+	derived, err := DeriveNonReversed(*s, DeriveReasonPUKPrev)
+	if err != nil {
+		return res, err
+	}
+	return NaclSecretBoxKey(derived), err
+}
+
+func (s *PerUserKeySeed) IsBlank() bool {
```

Comment:
Would it be safer to pass the seed around as a pointer, so that the default value is less likely to make it into any actual computation? (I'm not sure it matters, since slicing works on nil? Just thinking out loud.)